syntax = "proto3";

package api;

import "struct.proto";

option go_package = ".;api";

service CreateNotifications {
  rpc CreateNotificationsAction (NotificationCreateRequest) returns (NotificationCreateResponse);
  rpc CreateNotificationForUsers (NotificationCreateManualRequest) returns (NotificationCreateResponse);

  rpc GetNotifications (UserNotificationsRequest) returns (UserNotificationsResponse);
  rpc GetMassNotifications (UserMassNotificationRequest) returns (UserMassNotificationResponse);

  rpc MarkAsReadNotifications (NotificationManageRequest) returns (NotificationManageResponse);
  rpc DeleteNotifications (NotificationManageRequest) returns (NotificationManageResponse);

}

message NotificationCreateRequest {
  string action = 1;
  optional uint64 target_id = 2;
  optional uint32 target_type = 3;
  bool important = 4;

  int32 type = 5;

  optional string link = 7;
  optional string image = 8;
  optional string text = 9;
}

message NotificationCreateManualRequest {
  repeated int32 user_ids = 1;
  string settings_key = 2;

  optional uint64 target_id = 3;
  optional uint32 target_type = 4;
  string text = 5;
  bool text_as_model = 6;
  bool important = 7;
  bool confirmation = 8;

  optional string link = 9;
  optional string image = 10;
}

message NotificationCreateResponse {
  bool is_created = 1;
}

message Notification {
  int32 id = 1;

  google.protobuf.Struct target = 2;
  string target_type = 3;

  string image = 4;
  string link = 5;
  string text = 6;
  string date = 7;
  bool important = 8;

  bool confirmation = 9;
  bool read = 10;
}

message UserNotificationsRequest {
  int32 user_id = 1;
  int32 type = 2;

  optional bool only_important = 3;
  optional bool read = 4;

  int32 page = 5;
  int32 count = 6;
}

message UserNotificationsResponse {
  repeated Notification notifications = 1;
}


message UserMassNotification {
  google.protobuf.Struct notification = 1;

  bool read = 2;
}

message UserMassNotificationRequest {
  int32 type = 1;
  int32 user_id = 2;
  optional string action = 3;

  optional bool only_important = 4;
  optional bool read = 5;

  int32 page = 6;
  int32 count = 7;
}

message UserMassNotificationResponse {
  repeated UserMassNotification notifications = 1;
}

message NotificationManageRequest {
  repeated int32 notification_ids = 1;
  int32 user_id = 2;
}

message NotificationManageResponse {
  bool success = 1;
}